Src/FluentAssertions/Equivalency/SelfReferenceEquivalencyAssertionOptions.cs

```diff
@@ -182,7 +182,21 @@ EqualityStrategy IEquivalencyAssertionOptions.GetEqualityStrategy(Type requested
     {
         // As the valueFactory parameter captures instance members,
         // be aware if the cache must be cleared on mutating the members.
-        return equalityStrategyCache.GetOrAdd(requestedType, type =>
+#pragma warning disable SA1009, SA1111
+        return equalityStrategyCache.GetOrAdd(requestedType,
+#if NET47 || NETSTANDARD2_0
+            type => Add(type, referenceTypes, valueTypes, compareRecordsByValue, getDefaultEqualityStrategy)
+#else
+            static (Type type, (List<Type> referenceTypes, List<Type> valueTypes, bool compareRecordsByValue,
```

Comment:
Why do you need this?

Reply:
The idea was to use [`GetOrAdd<TArg> (TKey key, Func<TKey, TArg, TValue> valueFactory, TArg factoryArgument)`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.getoradd?view=net-6.0#system-collections-concurrent-concurrentdictionary-2-getoradd-1(-0-system-func((-0-0-1))-0)) 
and pass in the state as the `factoryArgument` to avoid creating a capturing lambda expression.
I did some more benchmarks on `BeEquivalentToWithDeeplyNestedStructures` and it showed a measurable negative effect on the runtime.
As the code is more complex and the memory savings are only moderate, I've reverted that commit.


develop	
|         Method |              Runtime |  N | Depth |      Mean |     Error |    StdDev |    Gen 0 |  Gen 1 | Allocated |
|--------------- |--------------------- |--- |------ |----------:|----------:|----------:|---------:|-------:|----------:|
| BeEquivalentTo |             .NET 6.0 |  1 |     1 |  32.98 us |  0.623 us |  0.692 us |   5.9204 | 0.1221 |     36 KB |
| BeEquivalentTo | .NET Framework 4.7.2 |  1 |     1 |  51.94 us |  0.378 us |  0.354 us |   6.4087 | 0.1221 |     39 KB |
| BeEquivalentTo |             .NET 6.0 |  1 |     2 |  41.43 us |  0.686 us |  0.642 us |   7.5684 | 0.1831 |     46 KB |
| BeEquivalentTo | .NET Framework 4.7.2 |  1 |     2 |  62.86 us |  0.205 us |  0.191 us |   8.1787 | 0.1221 |     51 KB |
| BeEquivalentTo |             .NET 6.0 |  1 |     6 |  77.23 us |  0.749 us |  0.701 us |  14.2822 | 0.4883 |     88 KB |
| BeEquivalentTo | .NET Framework 4.7.2 |  1 |     6 | 110.89 us |  0.493 us |  0.461 us |  15.5029 | 0.4883 |     96 KB |
| BeEquivalentTo |             .NET 6.0 | 10 |     1 | 204.92 us |  1.006 us |  0.941 us |  38.5742 | 1.4648 |    237 KB |
| BeEquivalentTo | .NET Framework 4.7.2 | 10 |     1 | 281.62 us |  3.471 us |  3.247 us |  41.9922 | 1.4648 |    258 KB |
| BeEquivalentTo |             .NET 6.0 | 10 |     2 | 291.63 us |  1.432 us |  1.340 us |  54.6875 | 1.9531 |    337 KB |
| BeEquivalentTo | .NET Framework 4.7.2 | 10 |     2 | 396.12 us |  6.709 us |  5.602 us |  59.5703 | 1.9531 |    368 KB |
| BeEquivalentTo |             .NET 6.0 | 10 |     6 | 629.06 us |  3.224 us |  3.016 us | 122.0703 | 6.8359 |    748 KB |
| BeEquivalentTo | .NET Framework 4.7.2 | 10 |     6 | 884.21 us | 11.413 us | 10.676 us | 132.8125 | 7.8125 |    818 KB |	


373a6add033fec1baf7870fcd10ba4ce61be2e44 - Lazily construct the equivalency options summary
|         Method |              Runtime |  N | Depth |      Mean |    Error |   StdDev |    Gen 0 |  Gen 1 | Allocated |
|--------------- |--------------------- |--- |------ |----------:|---------:|---------:|---------:|-------:|----------:|
| BeEquivalentTo |             .NET 6.0 |  1 |     1 |  31.98 us | 0.206 us | 0.193 us |   5.5542 | 0.1221 |     34 KB |
| BeEquivalentTo | .NET Framework 4.7.2 |  1 |     1 |  49.69 us | 0.961 us | 0.852 us |   6.0425 | 0.1221 |     37 KB |
| BeEquivalentTo |             .NET 6.0 |  1 |     2 |  41.15 us | 0.427 us | 0.400 us |   7.2021 | 0.1831 |     44 KB |
| BeEquivalentTo | .NET Framework 4.7.2 |  1 |     2 |  62.28 us | 0.495 us | 0.463 us |   7.8125 | 0.1221 |     48 KB |
| BeEquivalentTo |             .NET 6.0 |  1 |     6 |  75.42 us | 0.317 us | 0.296 us |  13.9160 | 0.4883 |     86 KB |
| BeEquivalentTo | .NET Framework 4.7.2 |  1 |     6 | 109.53 us | 0.823 us | 0.643 us |  15.1367 | 0.3662 |     94 KB |
| BeEquivalentTo |             .NET 6.0 | 10 |     1 | 202.09 us | 0.848 us | 0.752 us |  38.3301 | 1.2207 |    235 KB |
| BeEquivalentTo | .NET Framework 4.7.2 | 10 |     1 | 277.09 us | 2.755 us | 2.577 us |  41.5039 | 0.9766 |    256 KB |
| BeEquivalentTo |             .NET 6.0 | 10 |     2 | 287.58 us | 3.306 us | 3.092 us |  54.1992 | 1.9531 |    334 KB |
| BeEquivalentTo | .NET Framework 4.7.2 | 10 |     2 | 393.33 us | 3.889 us | 3.637 us |  59.0820 | 1.9531 |    365 KB |
| BeEquivalentTo |             .NET 6.0 | 10 |     6 | 653.00 us | 3.513 us | 2.934 us | 121.0938 | 6.8359 |    746 KB |
| BeEquivalentTo | .NET Framework 4.7.2 | 10 |     6 | 862.93 us | 9.103 us | 8.515 us | 131.8359 | 6.8359 |    816 KB |

6c5db114828617e2047588bea7750705d18e3893 - Avoid repeatedly constructing lambda
|         Method |              Runtime |  N | Depth |      Mean |     Error |   StdDev |    Gen 0 |  Gen 1 | Allocated |
|--------------- |--------------------- |--- |------ |----------:|----------:|---------:|---------:|-------:|----------:|
| BeEquivalentTo |             .NET 6.0 |  1 |     1 |  31.32 us |  0.404 us | 0.378 us |   5.1270 | 0.1221 |     32 KB |
| BeEquivalentTo | .NET Framework 4.7.2 |  1 |     1 |  50.09 us |  0.462 us | 0.409 us |   5.6152 | 0.1221 |     35 KB |
| BeEquivalentTo |             .NET 6.0 |  1 |     2 |  40.81 us |  0.412 us | 0.385 us |   6.6528 | 0.1831 |     41 KB |
| BeEquivalentTo | .NET Framework 4.7.2 |  1 |     2 |  60.81 us |  0.487 us | 0.407 us |   7.2632 | 0.1831 |     45 KB |
| BeEquivalentTo |             .NET 6.0 |  1 |     6 |  73.71 us |  0.394 us | 0.368 us |  12.6953 | 0.3662 |     78 KB |
| BeEquivalentTo | .NET Framework 4.7.2 |  1 |     6 | 108.38 us |  1.447 us | 1.354 us |  13.9160 | 0.3662 |     86 KB |
| BeEquivalentTo |             .NET 6.0 | 10 |     1 | 199.46 us |  0.758 us | 0.672 us |  34.9121 | 1.2207 |    214 KB |
| BeEquivalentTo | .NET Framework 4.7.2 | 10 |     1 | 277.28 us |  2.033 us | 1.901 us |  38.0859 | 0.9766 |    235 KB |
| BeEquivalentTo |             .NET 6.0 | 10 |     2 | 276.10 us |  2.801 us | 2.339 us |  49.3164 | 1.9531 |    303 KB |
| BeEquivalentTo | .NET Framework 4.7.2 | 10 |     2 | 395.81 us |  3.764 us | 3.521 us |  54.1992 | 1.9531 |    334 KB |
| BeEquivalentTo |             .NET 6.0 | 10 |     6 | 627.73 us |  5.543 us | 4.914 us | 109.3750 | 5.8594 |    675 KB |
| BeEquivalentTo | .NET Framework 4.7.2 | 10 |     6 | 858.80 us | 10.349 us | 9.680 us | 121.0938 | 6.8359 |    744 KB |


Avoid capture in GetEqualityStrategy
|         Method |              Runtime |  N | Depth |      Mean |    Error |   StdDev |    Gen 0 |  Gen 1 | Allocated |
|--------------- |--------------------- |--- |------ |----------:|---------:|---------:|---------:|-------:|----------:|
| BeEquivalentTo |             .NET 6.0 |  1 |     1 |  31.75 us | 0.179 us | 0.168 us |   5.0049 | 0.1221 |     31 KB |
| BeEquivalentTo | .NET Framework 4.7.2 |  1 |     1 |  49.64 us | 0.746 us | 0.698 us |   5.6152 | 0.1221 |     35 KB |
| BeEquivalentTo |             .NET 6.0 |  1 |     2 |  41.73 us | 0.385 us | 0.360 us |   6.4697 | 0.1221 |     40 KB |
| BeEquivalentTo | .NET Framework 4.7.2 |  1 |     2 |  60.87 us | 0.661 us | 0.586 us |   7.2021 | 0.1221 |     45 KB |
| BeEquivalentTo |             .NET 6.0 |  1 |     6 |  75.15 us | 0.192 us | 0.171 us |  12.3291 | 0.3662 |     76 KB |
| BeEquivalentTo | .NET Framework 4.7.2 |  1 |     6 | 109.04 us | 1.774 us | 1.385 us |  13.9160 | 0.3662 |     86 KB |
| BeEquivalentTo |             .NET 6.0 | 10 |     1 | 202.90 us | 1.279 us | 1.197 us |  33.9355 | 1.2207 |    208 KB |
| BeEquivalentTo | .NET Framework 4.7.2 | 10 |     1 | 284.56 us | 2.712 us | 2.537 us |  38.0859 | 0.9766 |    235 KB |
| BeEquivalentTo |             .NET 6.0 | 10 |     2 | 292.26 us | 1.784 us | 1.581 us |  47.8516 | 1.9531 |    295 KB |
| BeEquivalentTo | .NET Framework 4.7.2 | 10 |     2 | 399.16 us | 4.039 us | 3.778 us |  54.1992 | 1.9531 |    334 KB |
| BeEquivalentTo |             .NET 6.0 | 10 |     6 | 643.02 us | 2.192 us | 1.943 us | 106.4453 | 5.8594 |    657 KB |
| BeEquivalentTo | .NET Framework 4.7.2 | 10 |     6 | 878.81 us | 8.633 us | 8.075 us | 121.0938 | 6.8359 |    744 KB |


981570b16bd56b23c7d1dbf7724612289157150b - Avoid allocations in BecauseOf
|         Method |              Runtime |  N | Depth |      Mean |    Error |   StdDev |    Gen 0 |  Gen 1 | Allocated |
|--------------- |--------------------- |--- |------ |----------:|---------:|---------:|---------:|-------:|----------:|
| BeEquivalentTo |             .NET 6.0 |  1 |     1 |  30.45 us | 0.243 us | 0.227 us |   4.7607 | 0.0610 |     29 KB |
| BeEquivalentTo | .NET Framework 4.7.2 |  1 |     1 |  49.10 us | 0.930 us | 0.824 us |   5.3711 | 0.1221 |     33 KB |
| BeEquivalentTo |             .NET 6.0 |  1 |     2 |  39.17 us | 0.352 us | 0.329 us |   6.1035 | 0.1221 |     38 KB |
| BeEquivalentTo | .NET Framework 4.7.2 |  1 |     2 |  60.24 us | 0.823 us | 0.729 us |   6.9580 | 0.1221 |     43 KB |
| BeEquivalentTo |             .NET 6.0 |  1 |     6 |  72.41 us | 0.417 us | 0.390 us |  11.5967 | 0.3662 |     71 KB |
| BeEquivalentTo | .NET Framework 4.7.2 |  1 |     6 | 107.78 us | 0.234 us | 0.195 us |  13.1836 | 0.3662 |     81 KB |
| BeEquivalentTo |             .NET 6.0 | 10 |     1 | 192.19 us | 1.045 us | 0.926 us |  31.7383 | 0.9766 |    195 KB |
| BeEquivalentTo | .NET Framework 4.7.2 | 10 |     1 | 280.39 us | 2.284 us | 2.136 us |  35.6445 | 0.9766 |    221 KB |
| BeEquivalentTo |             .NET 6.0 | 10 |     2 | 277.87 us | 1.155 us | 1.081 us |  44.4336 | 1.4648 |    274 KB |
| BeEquivalentTo | .NET Framework 4.7.2 | 10 |     2 | 390.35 us | 3.603 us | 3.370 us |  50.7813 | 1.4648 |    313 KB |
| BeEquivalentTo |             .NET 6.0 | 10 |     6 | 604.42 us | 6.993 us | 6.199 us |  98.6328 | 5.8594 |    608 KB |
| BeEquivalentTo | .NET Framework 4.7.2 | 10 |     6 | 858.32 us | 6.407 us | 5.993 us | 112.3047 | 5.8594 |    696 KB |


8da904fd441afc6c313a86cea52980e10ea09515 - Avoid allocations in ConversionSelector
|         Method |              Runtime |  N | Depth |      Mean |     Error |   StdDev |    Gen 0 |  Gen 1 | Allocated |
|--------------- |--------------------- |--- |------ |----------:|----------:|---------:|---------:|-------:|----------:|
| BeEquivalentTo |             .NET 6.0 |  1 |     1 |  29.98 us |  0.214 us | 0.200 us |   4.5776 | 0.0610 |     28 KB |
| BeEquivalentTo | .NET Framework 4.7.2 |  1 |     1 |  47.50 us |  0.629 us | 0.558 us |   5.1880 | 0.0610 |     32 KB |
| BeEquivalentTo |             .NET 6.0 |  1 |     2 |  39.67 us |  0.315 us | 0.295 us |   5.8594 | 0.1221 |     36 KB |
| BeEquivalentTo | .NET Framework 4.7.2 |  1 |     2 |  58.46 us |  0.445 us | 0.394 us |   6.6528 | 0.1221 |     41 KB |
| BeEquivalentTo |             .NET 6.0 |  1 |     6 |  70.57 us |  0.320 us | 0.300 us |  10.9863 | 0.3662 |     68 KB |
| BeEquivalentTo | .NET Framework 4.7.2 |  1 |     6 | 104.00 us |  0.939 us | 0.832 us |  12.5732 | 0.3662 |     78 KB |
| BeEquivalentTo |             .NET 6.0 | 10 |     1 | 187.72 us |  0.860 us | 0.805 us |  30.2734 | 0.9766 |    185 KB |
| BeEquivalentTo | .NET Framework 4.7.2 | 10 |     1 | 265.62 us |  2.762 us | 2.584 us |  34.1797 | 0.9766 |    212 KB |
| BeEquivalentTo |             .NET 6.0 | 10 |     2 | 264.87 us |  1.372 us | 1.283 us |  42.4805 | 1.4648 |    262 KB |
| BeEquivalentTo | .NET Framework 4.7.2 | 10 |     2 | 379.77 us |  4.020 us | 3.563 us |  48.8281 | 1.4648 |    300 KB |
| BeEquivalentTo |             .NET 6.0 | 10 |     6 | 579.72 us |  1.538 us | 1.438 us |  93.7500 | 4.8828 |    579 KB |
| BeEquivalentTo | .NET Framework 4.7.2 | 10 |     6 | 834.86 us | 10.682 us | 9.469 us | 108.3984 | 5.8594 |    666 KB |


370c674a6c5c82343a4a5c2452a3897a4aafc9a7 - Avoid/cache string allocations
|         Method |              Runtime |  N | Depth |      Mean |     Error |   StdDev |   Gen 0 |  Gen 1 | Allocated |
|--------------- |--------------------- |--- |------ |----------:|----------:|---------:|--------:|-------:|----------:|
| BeEquivalentTo |             .NET 6.0 |  1 |     1 |  28.93 us |  0.293 us | 0.274 us |  4.3945 | 0.0610 |     27 KB |
| BeEquivalentTo | .NET Framework 4.7.2 |  1 |     1 |  46.16 us |  0.377 us | 0.353 us |  4.9438 | 0.0610 |     31 KB |
| BeEquivalentTo |             .NET 6.0 |  1 |     2 |  37.23 us |  0.341 us | 0.319 us |  5.4932 | 0.1221 |     34 KB |
| BeEquivalentTo | .NET Framework 4.7.2 |  1 |     2 |  56.37 us |  0.399 us | 0.354 us |  6.2256 | 0.1221 |     39 KB |
| BeEquivalentTo |             .NET 6.0 |  1 |     6 |  69.50 us |  0.335 us | 0.313 us |  9.8877 | 0.2441 |     61 KB |
| BeEquivalentTo | .NET Framework 4.7.2 |  1 |     6 |  97.00 us |  1.051 us | 0.878 us | 11.3525 | 0.2441 |     70 KB |
| BeEquivalentTo |             .NET 6.0 | 10 |     1 | 176.22 us |  1.161 us | 1.086 us | 28.0762 | 0.9766 |    172 KB |
| BeEquivalentTo | .NET Framework 4.7.2 | 10 |     1 | 252.86 us |  1.836 us | 1.718 us | 31.7383 | 0.9766 |    196 KB |
| BeEquivalentTo |             .NET 6.0 | 10 |     2 | 254.10 us |  1.439 us | 1.346 us | 38.5742 | 1.4648 |    239 KB |
| BeEquivalentTo | .NET Framework 4.7.2 | 10 |     2 | 350.46 us |  3.729 us | 3.306 us | 44.4336 | 1.4648 |    274 KB |
| BeEquivalentTo |             .NET 6.0 | 10 |     6 | 548.07 us |  2.091 us | 1.956 us | 82.0313 | 4.8828 |    507 KB |
| BeEquivalentTo | .NET Framework 4.7.2 | 10 |     6 | 759.56 us | 11.574 us | 9.036 us | 94.7266 | 4.8828 |    586 KB |